<a href="https://colab.research.google.com/github/mdheying/356-Python/blob/main/Error_propagation_with_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Background
Every measurement made in the laboratory has a finite uncertainty associated with it.  Likewise, any property calculated from one of these measurement will have an uncertainty that has been propagated from the original measurements.  For instance, suppose we wish to calculate the area of a rectangle by measuring its sides with a ruler, and we obstain the following results:

x = 15.50 cm &plusmn; 0.40 cm
y = 30.40 cm &plusmn; 0.70 cm

The area can be quickly calculated to be 471.2 cm<sup>2</sup>, but what is the uncertainty in the area?  This question can be answered by techniques of error propagation.  To calculate the propagated error, we need to know:


1.   the function, Z
2.   the uncertainties (or 'error') involved in the associated measurements.

The error in the variable i, usually denoted as σ<sub>i</sub>, must be determined by experiment. In our area example, the function is Z(x,y)=xy and the uncertainties are σ<sub>x</sub>=0.4, σ<sub>y</sub> = 0.7.  You have already learned the rules for propagating errors of simple sums, differences, products and ratios.  However, you will sometime need to propagate the error of a complicated function.  In this case, you must use the generate error formula:



If there were three variables measured, there would be three terms in the square root, and so on.  We could do this calculation with just a paper and pencil (see the example in LabPal), but software such as python can make our work much more efficient.  Below is a guided example using the rectangle described above.

# Implementation
When we propagate error, we need to utilize python's symbolic abilities.

```
from sympy import *
x, y, Area, sx, sy, sArea = symbols('x, y, Area, sx, sy, sArea')
```

To begin, we brought in SymPy (in its entirety) and then defined all of our variables as symbols.

Next, we define our function - in this example, the area of a rectangle.


```
Area = x*y
```

Finally, we set up the general error formula.  Remember that we are taking advantage of python's ability to take partial derivatives so that we do not need to do them.  That means the following pattern would apply to any measurement with two variables, not just our rectangle example.


```
sArea = sqrt(sx**2*(diff(Area,x))**2 + sy**2*(diff(Area,y))**2)
```

After this line, we have two symbolic expressions, *Area*, and *sArea*.  We could choose to display the equations, but typically, we are more interested in their values.
Our last step, then, is to substitute the actual values into the equations.


```
Area_val = Area.subs({x:15.5, y:30.4})
sArea_val = sArea.subs({x:15.5, y:30.4, sx:0.4, sy:0.7})
```

The code block below has combined all of the steps above along with display commands.

In [ ]:
from sympy import *
x, y, Area, sx, sy, sArea = symbols('x, y, Area, sx, sy, sArea')

Area = x*y
sArea = sqrt(sx**2*(diff(Area,x))**2 + sy**2*(diff(Area,y))**2)

Area_val = Area.subs({x:15.5, y:30.4})
sArea_val = sArea.subs({x:15.5, y:30.4, sx:0.4, sy:0.7})

display(Area_val)
display(sArea_val)

471.200000000000

16.2968739333653

# Significant figures
In this course (CHEM 356) we will apply a specific form of significant figures, different from the general rules you have likely used in the past.


1.   Round σ<sub>Z</sub> to **2** significant figures.
2.   Round Z so that it has the same number of decimal places.

In our example, this will look like:


1.   16.29876... becomes 16
2.   471.2... becomes 471

This gives us our final result for the area, 471 &plusmn; 16 cm<sup>2</sup>

# Propagation 1 (LabPal assignment)
In this exercise, you will calculate the density and its uncertainty.
Remember that density (represented in LabPal as ρ), is found from ρ=m/V
Don't forget to treat your significant figures appropriately at the end!

In [ ]:
# Here are a few lines to get you started
from sympy import *
m, V, dens, sm, sV, sdens = symbols('m, V, dens, sm, sV, sdens')


# Propagation 2 (LabPal assignment)
In this exercise, the formula is more complicated.  Unlike previous examples, it would take a long time to apply the simple propagation rules you may have used in the past.  This is definitely a time to use python!
*Aside*: Does the formula look familiar? This is the final temperature of two objects of the same material, initially at different temperatures, in contact with each other.

*Aside*: Does the formula look familiar? This is the final temperature of two objects of the same material, initially at different temperatures, in contact with each other.

In [ ]:
# Here are a few lines to get you started
from sympy import *
m1, m2, T1, T2, T, sm1, sm2, sT1, sT2 = symbols('m1, m2, T1, T2, T, sm1, sm2, sT1, sT2')


# Propagation 3 (LabPal assignment)
One last time!  We end on a dilution calculation.  See how you do!

In [ ]:
# Here is one line to get you started...
from sympy import *